File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/LinkedList/DoublyLinkedList.java
import java.util.Objects;

class DoublyNode<T> {
    T data;
    DoublyNode<T> next;
    DoublyNode<T> previous;

    public DoublyNode(T data) {
        this.data = data;
        this.next = null;
        this.previous = null;
    }
}

public class DoublyLinkedList<T> {
    private DoublyNode<T> head;
    private DoublyNode<T> tail;
    private int count;

    public void addFirst(T data) {
        DoublyNode<T> newNode = new DoublyNode<>(data);
        if (head == null) {
            head = tail = newNode;
        } else {
            newNode.next = head;
            head.previous = newNode;
            head = newNode;
        }
        count++;
    }

    public void addLast(T data) {
        DoublyNode<T> newNode = new DoublyNode<>(data);
        if (head == null) {
            head = tail = newNode;
        } else {
            tail.next = newNode;
            newNode.previous = tail;
            tail = newNode;
        }
        count++;
    }

    public boolean remove(T data) {
        DoublyNode<T> current = head;
        while (current != null) {
            if (Objects.equals(current.data, data)) {
                if (current.previous == null) {
                    head = current.next;
                    if (head != null) {
                        head.previous = null;
                    } else {
                        tail = null;
                    }
                } else if (current.next == null) {
                    tail = current.previous;
                    tail.next = null;
                } else {
                    current.previous.next = current.next;
                    current.next.previous = current.previous;
                }
                count--;
                return true;
            }
            current = current.next;
        }
        return false;
    }
} 

File: Aio-Algorithms-And-Solutions/Algorithms/DataStructures/LinkedList/DoublyLinkedList.cs
public class DoublyNode<T>
{
    public T Data { get; set; }
    public DoublyNode<T>? Next { get; set; }
    public DoublyNode<T>? Previous { get; set; }

    public DoublyNode(T data)
    {
        Data = data;
        Next = null;
        Previous = null;
    }
}

public class DoublyLinkedList<T>
{
    private DoublyNode<T>? head;
    private DoublyNode<T>? tail;
    public int Count { get; private set; }

    public void AddFirst(T data)
    {
        DoublyNode<T> newNode = new DoublyNode<T>(data);
        if (head == null)
        {
            head = tail = newNode;
        }
        else
        {
            newNode.Next = head;
            head.Previous = newNode;
            head = newNode;
        }
        Count++;
    }

    public void AddLast(T data)
    {
        DoublyNode<T> newNode = new DoublyNode<T>(data);
        if (head == null)
        {
            head = tail = newNode;
        }
        else
        {
            tail!.Next = newNode;
            newNode.Previous = tail;
            tail = newNode;
        }
        Count++;
    }

    public bool Remove(T data)
    {
        DoublyNode<T>? current = head;

        while (current != null)
        {
            if (EqualityComparer<T>.Default.Equals(current.Data, data))
            {
                if (current.Previous == null)
                {
                    head = current.Next;
                    if (head != null)
                        head.Previous = null;
                    else
                        tail = null;
                }
                else if (current.Next == null)
                {
                    tail = current.Previous;
                    tail.Next = null;
                }
                else
                {
                    current.Previous.Next = current.Next;
                    current.Next.Previous = current.Previous;
                }
                Count--;
                return true;
            }
            current = current.Next;
        }
        return false;
    }
} 